mirgecom/viscous.py

```diff
@@ -0,0 +1,377 @@
+        The species diffusive flux vector, $\mathbf{J}_{\alpha}$
+    """
+    nspecies = cv.nspecies
+    transport = eos.transport_model()
+
+    grad_y = species_mass_fraction_gradient(discr, cv, grad_cv)
+    d = transport.species_diffusivity(eos, cv)
+
+    # TODO: Better way?
+    obj_ary = -make_obj_array([cv.mass*d[i]*grad_y[i] for i in range(nspecies)])
+    diffusive_flux = np.empty(shape=(nspecies, discr.dim), dtype=object)
+    for idx, v in enumerate(obj_ary):
+        diffusive_flux[idx] = v
+
+    return diffusive_flux
+
+
+def conductive_heat_flux(discr, eos, cv, grad_t):
+    r"""Compute the conductive heat flux, ($\mathbf{q}_{c}$).
+
+    The conductive heat flux is defined by:
+
+    .. math::
+
+        \mathbf{q}_{c} = -\kappa\nabla{T},
+
+    with thermal conductivity $\kappa$, and gas temperature $T$.
+
+    Parameters
+    ----------
+    eos: :class:`~mirgecom.eos.GasEOS`
+        A gas equation of state with a non-empty
+        :class:`~mirgecom.transport.TransportModel`
+    cv: :class:`~mirgecom.fluid.ConservedVars`
+        Fluid state
+    grad_t: numpy.ndarray
+        Gradient of the fluid temperature
+
+    Returns
+    -------
+    numpy.ndarray
+        The conductive heat flux vector
+    """
+    transport = eos.transport_model()
+    return -transport.thermal_conductivity(eos, cv)*grad_t
+
+
+def diffusive_heat_flux(discr, eos, cv, j):
+    r"""Compute the diffusive heat flux, ($\mathbf{q}_{d}$).
+
+    The diffusive heat flux is defined by:
+
+    .. math::
+
+        \mathbf{q}_{d} = \sum_{\alpha=1}^{\mathtt{Nspecies}}{h}_{\alpha}
+        \mathbf{J}_{\alpha},
+
+    with species specific enthalpy ${h}_{\alpha}$ and diffusive flux
+    ($\mathbf{J}_{\alpha}$) defined as:
+
+    .. math::
+
+        \mathbf{J}_{\alpha} = -\rho{d}_{\alpha}\nabla{Y}_{\alpha},
+
+    where ${Y}_{\alpha}$ is the vector of species mass fractions.
+
+    Parameters
+    ----------
+    eos: mirgecom.eos.GasEOS
+        A gas equation of state with a non-empty
+        :class:`~mirgecom.transport.TransportModel`
+    cv: :class:`~mirgecom.fluid.ConservedVars`
+        Fluid state
+    j: numpy.ndarray
+        The species diffusive flux vector
+
+    Returns
+    -------
+    numpy.ndarray
+        The total diffusive heath flux vector
+    """
+    nspec = cv.nspecies
+    if nspec > 0:
+        try:
+            h_alpha = eos.get_species_enthalpies(cv)
+            return sum(h_alpha[i]*j[i] for i in range(nspec))
+        except NotImplementedError:
+            return 0
+    return 0
+
+
+# TODO: We could easily make this more general (dv, grad_dv)
+# so that user can define arbitrary array containers for *dv*,
+# enabling support for other models and simultaneous eqns.
+def viscous_flux(discr, eos, cv, grad_cv, grad_t):
+    r"""Compute the viscous flux vectors.
+
+    The viscous fluxes are:
+
+    .. math::
+
+        \mathbf{F}_V = [0,\tau\cdot\mathbf{v} - \mathbf{q},
+        \tau_{:i},-\mathbf{J}_\alpha],
```

Comment:
What is `\tau_{:i}`?

Reply:
F_V = [ 0, \tau.dot.v - q, row 1 of \tau, .. row (dim) of \tau, -J]

I see this \tau{:i} notation used in the literature, but I'll see if I can make this doc more clear.

Reply:
Heh, well... see what you think of it now.

Reply:
I think this looks OK now. Missing a definition for `q` though.